File: binary-trees/convert-bst-to-greater-tree/convert-bst-to-greater-tree.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
  public TreeNode convertBST(TreeNode root) {
    traverse(root, 0);

    return root;
  }

  private int traverse(TreeNode root, int sum) {
    if (root == null) {
      return sum;
    }

    sum = traverse(root.right, sum);

    sum += root.val;

    root.val = sum;

    sum = traverse(root.left, sum);

    return sum;
  }
}

File: binary-trees/convert-bst-to-greater-tree/convert-bst-to-greater-tree.go
package convertbsttogreatertree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func convertBST(root *TreeNode) *TreeNode {
	traverse(root, 0)
	return root
}

func traverse(root *TreeNode, sum int) int {
	if root == nil {
		return sum
	}

	sum = traverse(root.Right, sum)
	sum += root.Val
	root.Val = sum
	sum = traverse(root.Left, sum)
	return sum
}
